File: georocket-server/src/main/java/io/georocket/ImporterVerticle.java
package io.georocket;

import io.georocket.constants.AddressConstants;
import io.georocket.constants.ConfigConstants;
import io.georocket.index.xml.XMLCRSIndexer;
import io.georocket.input.Splitter.Result;
import io.georocket.input.geojson.GeoJsonSplitter;
import io.georocket.input.xml.FirstLevelSplitter;
import io.georocket.input.xml.XMLSplitter;
import io.georocket.storage.ChunkMeta;
import io.georocket.storage.IndexMeta;
import io.georocket.storage.RxStore;
import io.georocket.storage.StoreFactory;
import io.georocket.tasks.ImportingTask;
import io.georocket.tasks.TaskError;
import io.georocket.util.JsonParserTransformer;
import io.georocket.util.RxUtils;
import io.georocket.util.StringWindow;
import io.georocket.util.UTF8BomFilter;
import io.georocket.util.Window;
import io.georocket.util.XMLParserTransformer;
import io.georocket.util.io.RxGzipReadStream;
import io.vertx.core.file.OpenOptions;
import io.vertx.core.impl.NoStackTraceThrowable;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;
import io.vertx.core.logging.Logger;
import io.vertx.core.logging.LoggerFactory;
import io.vertx.rxjava.core.AbstractVerticle;
import io.vertx.rxjava.core.buffer.Buffer;
import io.vertx.rxjava.core.eventbus.Message;
import io.vertx.rxjava.core.file.AsyncFile;
import io.vertx.rxjava.core.file.FileSystem;
import io.vertx.rxjava.core.streams.ReadStream;
import rx.Completable;
import rx.Observable;
import rx.Single;

import java.time.Instant;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.Consumer;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static io.georocket.util.MimeTypeUtils.belongsTo;

/**
 * Imports file in the background
 * @author Michel Kraemer
 */
public class ImporterVerticle extends AbstractVerticle {
  private static Logger log = LoggerFactory.getLogger(ImporterVerticle.class);
  
  private static final int MAX_RETRIES = 5;
  private static final int RETRY_INTERVAL = 1000;
  private static final int MAX_PARALLEL_IMPORTS = 1;
  
  protected RxStore store;
  private String incoming;
  private boolean paused;
  private Set<AsyncFile> filesBeingImported = new HashSet<>();

  @Override
  public void start() {
    log.info("Launching importer ...");

    store = new RxStore(StoreFactory.createStore(getVertx()));
    String storagePath = config().getString(ConfigConstants.STORAGE_FILE_PATH);
    incoming = storagePath + "/incoming";
    
    vertx.eventBus().<JsonObject>localConsumer(AddressConstants.IMPORTER_IMPORT)
      .toObservable()
      .onBackpressureBuffer() // unlimited buffer
      .flatMapCompletable(msg -> {
        // call onImport() but ignore errors. onImport() will handle errors for us.
        return onImport(msg).onErrorComplete();
      }, false, MAX_PARALLEL_IMPORTS)
      .subscribe(v -> {
        // ignore
      }, err -> {
        // This is bad. It will unsubscribe the consumer from the eventbus!
        // Should never happen anyhow. If it does, something else has
        // completely gone wrong.
        log.fatal("Could not import file", err);
      });

    vertx.eventBus().localConsumer(AddressConstants.IMPORTER_PAUSE, this::onPause);
  }

  /**
   * Receives a name of a file to import
   * @param msg the event bus message containing the filename
   * @return a Completable that will complete when the file has been imported
   */
  protected Completable onImport(Message<JsonObject> msg) {
    JsonObject body = msg.body();
    String filename = body.getString("filename");
    String filepath = incoming + "/" + filename;
    String layer = body.getString("layer", "/");
    String contentType = body.getString("contentType");
    String correlationId = body.getString("correlationId");
    String fallbackCRSString = body.getString("fallbackCRSString");
    String contentEncoding = body.getString("contentEncoding");

    // get tags
    JsonArray tagsArr = body.getJsonArray("tags");
    List<String> tags = tagsArr != null ? tagsArr.stream().flatMap(o -> o != null ?
        Stream.of(o.toString()) : Stream.of()).collect(Collectors.toList()) : null;

    // get properties
    JsonObject propertiesObj = body.getJsonObject("properties");
    Map<String, Object> properties = propertiesObj != null ? propertiesObj.getMap() : null;

    // generate timestamp for this import
    long timestamp = System.currentTimeMillis();

    log.info("Importing [" + correlationId + "] to layer '" + layer + "'");

    FileSystem fs = vertx.fileSystem();
    OpenOptions openOptions = new OpenOptions().setCreate(false).setWrite(false);
    return fs.rxOpen(filepath, openOptions)
      .flatMap(f -> {
        filesBeingImported.add(f);
        return importFile(contentType, f, correlationId, filename, timestamp,
            layer, tags, properties, fallbackCRSString, contentEncoding)
          .doAfterTerminate(() -> {
            // delete file from 'incoming' folder
            log.debug("Deleting " + filepath + " from incoming folder");
            filesBeingImported.remove(f);
            f.rxClose()
              .flatMap(v -> fs.rxDelete(filepath))
              .subscribe(v -> {}, err -> {
                log.error("Could not delete file from 'incoming' folder", err);
              });
          });
      })
      .doOnSuccess(chunkCount -> {
        long duration = System.currentTimeMillis() - timestamp;
        log.info("Finished importing [" + correlationId + "] with " + chunkCount +
            " chunks to layer '" + layer + "' after " + duration + " ms");
      })
      .doOnError(err -> {
        long duration = System.currentTimeMillis() - timestamp;
        log.error("Failed to import [" + correlationId + "] to layer '" +
            layer + "' after " + duration + " ms", err);
      })
      .toCompletable();
  }

  /**
   * Import a file from the given read stream into the store. Inspect the file's
   * content type and forward to the correct import method.
   * @param contentType the file's content type
   * @param f the file to import
   * @param correlationId a unique identifier for this import process
   * @param filename the name of the file currently being imported
   * @param timestamp denotes when the import process has started
   * @param layer the layer where the file should be stored (may be null)
   * @param tags the list of tags to attach to the file (may be null)
   * @param properties the map of properties to attach to the file (may be null)
   * @param fallbackCRSString the CRS which should be used if the imported
   * file does not specify one (may be <code>null</code>)
   * @param contentEncoding the content encoding of the file to be
   * imported (e.g. "gzip"). May be <code>null</code>.
   * @return a single that will emit with the number if chunks imported
   * when the file has been imported
   */
  protected Single<Integer> importFile(String contentType, ReadStream<Buffer> f,
      String correlationId, String filename, long timestamp, String layer,
      List<String> tags, Map<String, Object> properties, String fallbackCRSString,
      String contentEncoding) {
    if ("gzip".equals(contentEncoding)) {
      log.debug("Importing file compressed with GZIP");
      f = new RxGzipReadStream(f);
    } else if (contentEncoding != null && !contentEncoding.isEmpty()) {
      log.warn("Unknown content encoding: `" + contentEncoding + "'. Trying anyway.");
    }

    // let the task verticle know that we're now importing
    ImportingTask startTask = new ImportingTask(correlationId);
    startTask.setStartTime(Instant.now());
    vertx.eventBus().publish(AddressConstants.TASK_INC, JsonObject.mapFrom(startTask));

    Observable<Integer> result;
    if (belongsTo(contentType, "application", "xml") ||
        belongsTo(contentType, "text", "xml")) {
      result = importXML(f, correlationId, filename, timestamp, layer, tags,
        properties, fallbackCRSString);
    } else if (belongsTo(contentType, "application", "json")) {
      result = importJSON(f, correlationId, filename, timestamp, layer, tags, properties);
    } else {
      result = Observable.error(new NoStackTraceThrowable(String.format(
          "Received an unexpected content type '%s' while trying to import "
          + "file '%s'", contentType, filename)));
    }

    Consumer<Throwable> onFinish = t -> {
      // let the task verticle know that the import process has finished
      ImportingTask endTask = new ImportingTask(correlationId);
      endTask.setEndTime(Instant.now());
      if (t != null) {
        endTask.addError(new TaskError(t));
      }
      vertx.eventBus().publish(AddressConstants.TASK_INC,
          JsonObject.mapFrom(endTask));
    };

    return result.window(100)
      .flatMap(Observable::count)
      .doOnNext(n -> {
        // let the task verticle know that we imported n chunks
        ImportingTask currentTask = new ImportingTask(correlationId);
        currentTask.setImportedChunks(n);
        vertx.eventBus().publish(AddressConstants.TASK_INC,
            JsonObject.mapFrom(currentTask));
      })
      .reduce(0, (a, b) -> a + b)
      .toSingle()
      .doOnError(onFinish::accept)
      .doOnSuccess(i -> onFinish.accept(null));
  }

  /**
   * Imports an XML file from the given input stream into the store
   * @param f the XML file to read
   * @param correlationId a unique identifier for this import process
   * @param filename the name of the file currently being imported
   * @param timestamp denotes when the import process has started
   * @param layer the layer where the file should be stored (may be null)
   * @param tags the list of tags to attach to the file (may be null)
   * @param properties the map of properties to attach to the file (may be null)
   * @param fallbackCRSString the CRS which should be used if the imported
   * file does not specify one (may be <code>null</code>)
   * @return an observable that will emit the number 1 when a chunk has been imported
   */
  protected Observable<Integer> importXML(ReadStream<Buffer> f, String correlationId,
      String filename, long timestamp, String layer, List<String> tags,
      Map<String, Object> properties, String fallbackCRSString) {
    UTF8BomFilter bomFilter = new UTF8BomFilter();
    Window window = new Window();
    XMLSplitter splitter = new FirstLevelSplitter(window);
    AtomicInteger processing = new AtomicInteger(0);
    XMLCRSIndexer crsIndexer = new XMLCRSIndexer();
    return f.toObservable()
        .map(Buffer::getDelegate)
        .map(bomFilter::filter)
        .doOnNext(window::append)
        .compose(new XMLParserTransformer())
        .doOnNext(e -> {
          // save the first CRS found in the file
          if (crsIndexer.getCRS() == null) {
            crsIndexer.onEvent(e);
          }
        })
        .flatMap(splitter::onEventObservable)
        .flatMapSingle(result -> {
          String crsString = fallbackCRSString;
          if (crsIndexer.getCRS() != null) {
            crsString = crsIndexer.getCRS();
          }
          IndexMeta indexMeta = new IndexMeta(correlationId, filename,
              timestamp, tags, properties, crsString);
          return addToStoreWithPause(result, layer, indexMeta, f, processing)
              .toSingleDefault(1);
        });
  }
  
  /**
   * Imports a JSON file from the given input stream into the store
   * @param f the JSON file to read
   * @param correlationId a unique identifier for this import process
   * @param filename the name of the file currently being imported
   * @param timestamp denotes when the import process has started
   * @param layer the layer where the file should be stored (may be null)
   * @param tags the list of tags to attach to the file (may be null)
   * @param properties the map of properties to attach to the file (may be null)
   * @return an observable that will emit the number 1 when a chunk has been imported
   */
  protected Observable<Integer> importJSON(ReadStream<Buffer> f, String correlationId,
      String filename, long timestamp, String layer, List<String> tags, Map<String, Object> properties) {
    UTF8BomFilter bomFilter = new UTF8BomFilter();
    StringWindow window = new StringWindow();
    GeoJsonSplitter splitter = new GeoJsonSplitter(window);
    AtomicInteger processing = new AtomicInteger(0);
    return f.toObservable()
        .map(Buffer::getDelegate)
        .map(bomFilter::filter)
        .doOnNext(window::append)
        .compose(new JsonParserTransformer())
        .flatMap(splitter::onEventObservable)
        .flatMapSingle(result -> {
          IndexMeta indexMeta = new IndexMeta(correlationId, filename,
              timestamp, tags, properties, null);
          return addToStoreWithPause(result, layer, indexMeta, f, processing)
              .toSingleDefault(1);
        });
  }

  /**
   * Handle a pause message
   * @param msg the message
   */
  private void onPause(Message<Boolean> msg) {
    Boolean paused = msg.body();
    if (paused == null || !paused) {
      if (this.paused) {
        log.info("Resuming import");
        this.paused = false;
        for (AsyncFile f : filesBeingImported) {
          f.resume();
        }
      }
    } else {
      if (!this.paused) {
        log.info("Pausing import");
        this.paused = true;
        for (AsyncFile f : filesBeingImported) {
          f.pause();
        }
      }
    }
  }

  /**
   * Add a chunk to the store. Pause the given read stream before adding and
   * increase the given counter. Decrease the counter after the chunk has been
   * written and only resume the read stream if the counter is <code>0</code>.
   * This is necessary because the writing to the store may take longer than
   * reading. We need to pause reading so the store is not overloaded (i.e.
   * we handle back-pressure here).
   * @param chunk the chunk to write
   * @param layer the layer the chunk should be added to (may be null)
   * @param indexMeta metadata specifying how the chunk should be indexed
   * @param f the read stream to pause while writing
   * @param processing an AtomicInteger keeping the number of chunks currently
   * being written (should be initialized to <code>0</code> the first time this
   * method is called)
   * @return a Completable that will complete when the operation has finished
   */
  private Completable addToStoreWithPause(Result<? extends ChunkMeta> chunk,
      String layer, IndexMeta indexMeta, ReadStream<Buffer> f, AtomicInteger processing) {
    // pause stream while chunk is being written
    f.pause();
    
    // count number of chunks being written
    processing.incrementAndGet();

    return addToStore(chunk.getChunk(), chunk.getMeta(), layer, indexMeta)
        .doOnCompleted(() -> {
          // resume stream only after all chunks from the current
          // buffer have been stored
          if (processing.decrementAndGet() == 0 && !paused) {
            f.resume();
          }
        });
  }
  
  /**
   * Add a chunk to the store. Retry operation several times before failing.
   * @param chunk the chunk to add
   * @param meta the chunk's metadata
   * @param layer the layer the chunk should be added to (may be null)
   * @param indexMeta metadata specifying how the chunk should be indexed
   * @return a Completable that will complete when the operation has finished
   */
  protected Completable addToStore(String chunk, ChunkMeta meta,
      String layer, IndexMeta indexMeta) {
    return Completable.defer(() -> store.rxAdd(chunk, meta, layer, indexMeta))
        .retryWhen(RxUtils.makeRetry(MAX_RETRIES, RETRY_INTERVAL, log));
  }
}

File: georocket-server/src/main/kotlin/io/georocket/ImporterVerticle.kt
package io.georocket

import com.fasterxml.aalto.AsyncXMLStreamReader
import com.fasterxml.aalto.stax.InputFactoryImpl
import de.undercouch.actson.JsonEvent
import de.undercouch.actson.JsonParser
import io.georocket.constants.AddressConstants
import io.georocket.constants.ConfigConstants
import io.georocket.index.xml.XMLCRSIndexer
import io.georocket.input.geojson.GeoJsonSplitter
import io.georocket.input.xml.FirstLevelSplitter
import io.georocket.storage.IndexMeta
import io.georocket.storage.Store
import io.georocket.storage.StoreFactory
import io.georocket.tasks.ImportingTask
import io.georocket.tasks.TaskError
import io.georocket.util.JsonStreamEvent
import io.georocket.util.MimeTypeUtils.belongsTo
import io.georocket.util.StringWindow
import io.georocket.util.UTF8BomFilter
import io.georocket.util.Window
import io.georocket.util.XMLStreamEvent
import io.georocket.util.io.GzipReadStream
import io.vertx.core.buffer.Buffer
import io.vertx.core.eventbus.Message
import io.vertx.core.json.JsonObject
import io.vertx.core.logging.LoggerFactory
import io.vertx.core.streams.ReadStream
import io.vertx.kotlin.core.file.closeAwait
import io.vertx.kotlin.core.file.deleteAwait
import io.vertx.kotlin.core.file.openAwait
import io.vertx.kotlin.core.file.openOptionsOf
import io.vertx.kotlin.coroutines.CoroutineVerticle
import io.vertx.kotlin.coroutines.toChannel
import kotlinx.coroutines.launch
import java.time.Instant

/**
 * Imports file in the background
 * @author Michel Kraemer
 */
class ImporterVerticle : CoroutineVerticle() {
  companion object {
    private val log = LoggerFactory.getLogger(ImporterVerticle::class.java)
  }

  private lateinit var store: Store
  private lateinit var incoming: String

  override suspend fun start() {
    log.info("Launching importer ...")

    store = StoreFactory.createStore(vertx)
    val storagePath = config.getString(ConfigConstants.STORAGE_FILE_PATH)
    incoming = "$storagePath/incoming"

    vertx.eventBus().localConsumer<JsonObject>(AddressConstants.IMPORTER_IMPORT) { msg ->
      launch {
        onImport(msg)
      }
    }
  }

  override suspend fun stop() {
    store.close()
  }

  /**
   * Receives a name of a file to import
   */
  private suspend fun onImport(msg: Message<JsonObject>) {
    val body = msg.body()
    val filename = body.getString("filename")
    val filepath = "$incoming/$filename"
    val layer = body.getString("layer", "/")
    val contentType = body.getString("contentType")
    val correlationId = body.getString("correlationId")
    val fallbackCRSString = body.getString("fallbackCRSString")
    val contentEncoding = body.getString("contentEncoding")

    // get tags
    val tags = body.getJsonArray("tags")
      ?.filterNotNull()
      ?.map { it.toString() }
      ?.toList()

    // get properties
    val properties = body.getJsonObject("properties")?.map

    // generate timestamp for this import
    val timestamp = System.currentTimeMillis()

    log.info("Importing [$correlationId] to layer '$layer'")

    val fs = vertx.fileSystem()
    val f = fs.openAwait(filepath, openOptionsOf(create = false, write = false))

    try {
      val chunkCount = importFile(f, contentType, correlationId, filename,
          timestamp, layer, tags, properties, fallbackCRSString, contentEncoding)

      val duration = System.currentTimeMillis() - timestamp
      log.info("Finished importing [$correlationId] with $chunkCount" +
          " chunks to layer '$layer' after $duration ms")
    } catch (t: Throwable) {
      val duration = System.currentTimeMillis() - timestamp
      log.error("Failed to import [$correlationId] to layer '$layer'" +
          " after $duration ms", t)
    } finally {
      // delete file from 'incoming' folder
      log.debug("Deleting $filepath from incoming folder")
      f.closeAwait()
      try {
        fs.deleteAwait(filepath)
      } catch (t: Throwable) {
        log.error("Could not delete file from 'incoming' folder", t)
      }
    }
  }

  /**
   * Import a [file] from the given read stream into the store. Inspect the
   * file's [contentType] and forward to the correct import method. Callers
   * should provide a [correlationId] so the import progress can be tracked
   * correctly. The method will also attach metadata to the imported chunks
   * such as the original [filename], the [timestamp] when the import was
   * started, the [layer] where chunks will be stored, as well as optional
   * [tags] and [properties]. If necessary, a [fallbackCRSString] can be given
   * if the imported file does not specify one. The file will be read as a raw
   * stream or using the given [contentEncoding] if provided.
   */
  private suspend fun importFile(file: ReadStream<Buffer>, contentType: String,
      correlationId: String, filename: String, timestamp: Long, layer: String,
      tags: List<String>?, properties: Map<String, Any>?, fallbackCRSString: String?,
      contentEncoding: String?): Long {
    val f = if ("gzip" == contentEncoding) {
      log.debug("Importing file compressed with GZIP")
      GzipReadStream(file)
    } else if (contentEncoding != null && contentEncoding.isNotEmpty()) {
      log.warn("Unknown content encoding: `$contentEncoding'. Trying anyway.")
      file
    } else {
      file
    }

    // let the task verticle know that we're now importing
    val startTask = ImportingTask(correlationId)
    startTask.startTime = Instant.now()
    vertx.eventBus().publish(AddressConstants.TASK_INC, JsonObject.mapFrom(startTask))

    var lastProgress = 0L
    val progressUpdater = { progress: Long, final: Boolean ->
      val diff = progress - lastProgress
      if (diff >= 100 || final) {
        // let the task verticle know that we imported n chunks
        val currentTask = ImportingTask(correlationId)
        currentTask.importedChunks = diff
        vertx.eventBus().publish(AddressConstants.TASK_INC,
          JsonObject.mapFrom(currentTask))
        lastProgress = progress
      }
    }

    try {
      val result = if (belongsTo(contentType, "application", "xml") ||
          belongsTo(contentType, "text", "xml")) {
        importXML(f, correlationId, filename, timestamp, layer, tags,
            properties, fallbackCRSString, progressUpdater)
      } else if (belongsTo(contentType, "application", "json")) {
        importJSON(f, correlationId, filename, timestamp, layer, tags,
          properties, progressUpdater)
      } else {
        throw IllegalArgumentException("Received an unexpected content " +
            "type '$contentType' while trying to import file '$filename'")
      }

      // let the task verticle know that the import process has finished
      val endTask = ImportingTask(correlationId)
      endTask.endTime = Instant.now()
      vertx.eventBus().publish(AddressConstants.TASK_INC,
          JsonObject.mapFrom(endTask))

      return result
    } catch (t: Throwable) {
      val endTask = ImportingTask(correlationId)
      endTask.endTime = Instant.now()
      endTask.addError(TaskError(t))
      vertx.eventBus().publish(AddressConstants.TASK_INC,
          JsonObject.mapFrom(endTask))
      throw t
    }
  }

  /**
   * Imports an XML file from the given input stream into the store
   * @param f the XML file to read
   * @param correlationId a unique identifier for this import process
   * @param filename the name of the file currently being imported
   * @param timestamp denotes when the import process has started
   * @param layer the layer where the file should be stored
   * @param tags the list of tags to attach to the file
   * @param properties the map of properties to attach to the file
   * @param fallbackCRSString the CRS which should be used if the imported
   * file does not specify one
   * @return an observable that will emit the number 1 when a chunk has been imported
   */
  private suspend fun importXML(f: ReadStream<Buffer>, correlationId: String,
      filename: String, timestamp: Long, layer: String, tags: List<String>?,
      properties: Map<String, Any>?, fallbackCRSString: String?,
      updateProgress: (Long, Boolean) -> Unit): Long {
    var chunksAdded = 0L

    val bomFilter = UTF8BomFilter()
    val window = Window()
    val splitter = FirstLevelSplitter(window)
    val crsIndexer = XMLCRSIndexer()
    val parser = InputFactoryImpl().createAsyncForByteArray()

    val makeIndexMeta = { crsString: String? ->
      IndexMeta(correlationId, filename, timestamp, tags, properties, crsString)
    }

    var indexMeta = makeIndexMeta(fallbackCRSString)

    val processEvents: suspend () -> Boolean = pe@{
      while (true) {
        val nextEvent = parser.next()
        if (nextEvent == AsyncXMLStreamReader.EVENT_INCOMPLETE) {
          break
        }

        // create stream event
        val pos = parser.location.characterOffset
        val streamEvent = XMLStreamEvent(nextEvent, pos, parser)

        // save the first CRS found in the file
        if (crsIndexer.crs == null) {
          crsIndexer.onEvent(streamEvent)
          if (crsIndexer.crs != null) {
            indexMeta = makeIndexMeta(crsIndexer.crs)
          }
        }

        val result = splitter.onEvent(streamEvent)
        if (result != null) {
          store.add(result.chunk, result.meta, indexMeta, layer)
          chunksAdded++
          updateProgress(chunksAdded, false)
        }

        if (nextEvent == AsyncXMLStreamReader.END_DOCUMENT) {
          parser.close()
          return@pe false
        }
      }
      true
    }

    val channel = f.toChannel(vertx)
    for (bytebuf in channel) {
      val buf = bomFilter.filter(bytebuf)
      window.append(buf)

      val bytes = buf.bytes
      var i = 0
      while (i < bytes.size) {
        val len = bytes.size - i
        parser.inputFeeder.feedInput(bytes, i, len)
        i += len
        if (!processEvents()) {
          break
        }
      }
    }

    // process remaining events
    parser.inputFeeder.endOfInput()
    processEvents()

    updateProgress(chunksAdded, true)

    return chunksAdded
  }

  /**
   * Imports a JSON file from the given input stream into the store
   * @param f the JSON file to read
   * @param correlationId a unique identifier for this import process
   * @param filename the name of the file currently being imported
   * @param timestamp denotes when the import process has started
   * @param layer the layer where the file should be stored
   * @param tags the list of tags to attach to the file
   * @param properties the map of properties to attach to the file
   * @return an observable that will emit the number 1 when a chunk has been imported
   */
  private suspend fun importJSON(f: ReadStream<Buffer>, correlationId: String,
      filename: String, timestamp: Long, layer: String, tags: List<String>?,
      properties: Map<String, Any>?, updateProgress: (Long, Boolean) -> Unit): Long {
    var chunksAdded = 0L

    val bomFilter = UTF8BomFilter()
    val window = StringWindow()
    val splitter = GeoJsonSplitter(window)
    val parser = JsonParser()

    val indexMeta = IndexMeta(correlationId, filename, timestamp, tags,
        properties, null)

    val processEvents: suspend () -> Boolean = pe@{
      while (true) {
        val nextEvent = parser.nextEvent()
        val value = when (nextEvent) {
          JsonEvent.NEED_MORE_INPUT -> break
          JsonEvent.ERROR -> throw IllegalStateException("Syntax error")
          JsonEvent.VALUE_STRING, JsonEvent.FIELD_NAME -> parser.currentString
          JsonEvent.VALUE_DOUBLE -> parser.currentDouble
          JsonEvent.VALUE_INT -> parser.currentInt
          else -> null
        }

        val streamEvent = JsonStreamEvent(nextEvent, parser.parsedCharacterCount, value)
        val result = splitter.onEvent(streamEvent)
        if (result != null) {
          store.add(result.chunk, result.meta, indexMeta, layer)
          chunksAdded++
          updateProgress(chunksAdded, false)
        }

        if (nextEvent == JsonEvent.EOF) {
          return@pe false
        }
      }
      true
    }

    val channel = f.toChannel(vertx)
    for (bytebuf in channel) {
      val buf = bomFilter.filter(bytebuf)
      window.append(buf)

      val bytes = buf.bytes
      var i = 0
      while (i < bytes.size) {
        i += parser.feeder.feed(bytes, i, bytes.size - i)
        if (!processEvents()) {
          break
        }
      }
    }

    // process remaining events
    parser.feeder.done()
    processEvents()

    updateProgress(chunksAdded, true)

    return chunksAdded
  }
}
